cmd/queue/main.go

```diff
@@ -137,6 +137,12 @@ func main() {
 		}.String()),
 		zap.String(logkey.Pod, env.ServingPod))
 
+	// initialize QPExtensions
+	for _, ext := range extensions {
+		ext.Init(logger, ctx)
+		defer ext.Shutdown()
+	}
```

Comment:
The only benefit I see of having `Shutdown` be part of the interface (rather than being part of the function outside `MainWithPlugins`) is that it may take advantage of the call to `flush()` on line 131.

The cost is that every implementation of `queue.QPExtension` must implement `Shutdown`. The alternative is to make `Flush` public, and ask callers which care to run `flush` again.

A third option is to have `MainWithPlugs()` return a `func()` for final cleanup which would return `func() {flush(logger)}`. This would have the advantage that it also worked with other post-`MainWithPlugs` code, with the disadvantage that you'd want to call `MainWithPlugs` as `defer MainWithPlugs(myhandlers...)()` or the like.

Reply:
yep, since `defer flush(logger)` is called first, and `defer ext.Shutdown()` after, 
`flush()` will be called after `Shutdown()` 

This seems like a clean way to go.